lib/tasks/data-key.rake

```diff
@@ -8,4 +8,63 @@ namespace :"data-key" do
     key = Base64.strict_encode64(Slosilo::Symmetric.new.random_key)
     print key
   end
+
+  desc "Validates the value of ENV['CONJUR_DATA_KEY']"
+  # This task validate the master data key which is stored in `ENV['CONJUR_DATA_KEY']`
+  # The task runs in two modes:
+  # First mode:
+  # ==========
+  #   Once the db is created, a role + resource + secret will be created using the master key.
+  #
+  # Second mode:
+  # ===========
+  #   Each time the server restarts this task will attempt to retrieve the
+  #   secret, if the provided key does not match the original master the secret retrieval
+  #   will fail and the task will exit with error.
+  #
+  # Side Note
+  # =========
+  #  What's the 'environment' task in Rake?
+  #  the :validate => :environment means "do environment before validate".
+  #  You can get access to your models, and in fact, your whole environment
+  #  by making tasks dependent on the environment task.
+  task :validate => [:environment] do
+    data_key = ENV['CONJUR_DATA_KEY']
+
+    if data_key.nil? || data_key.empty?
+      raise Errors::System::EmptyConjurDataKey
+      exit 1
+    end
+
+    # try get the master key validator resource
+    resource_id = "!:!:data-key-validator-resource"
+    if Resource[resource_id]
+      # Server restart - verify you can decrypt the secret
+      begin
+        # Attempt to retrieve the secret encrypt with the original data key.
+        # A failure to decrypt the value will indicate the key is invalid.
+        Resource[resource_id].secret.value
+      rescue => e
+        if e.inspect == 'OpenSSL::Cipher::CipherError'
+          raise Errors::System::InvalidConjurDataKey
+        else
+          $stderr.puts "Error retrieving data-key validator secret in Resource[#{resource_id}], Error: #{e.inspect}"
+          exit 1
+        end
+      end
+    else
+      # Server first run
+      # create the master key validator resource
+      # and store a secret using the master key
+      role_id = "!:!:data-key-validator"
+
+      begin
+        Role.create(role_id: role_id)
+        Resource.create(resource_id: resource_id, owner_id: role_id)
+        Secret.create(resource_id: resource_id, value: "Don't Panic.")
+      rescue => e
+        $stderr.puts "Error creating a resource for rake: data_key_validation, Error: #{e.inspect}"
```

Comment:
What is rake? I saw it was capitalized earlier. Should it be Rake here too?

Reply:
Rake is a software task management and build automation tool.